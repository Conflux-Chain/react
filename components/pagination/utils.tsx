export const getPageCount = (total: number, pageSize: number) => {
  return Math.floor((total - 1) / pageSize) + 1
}
