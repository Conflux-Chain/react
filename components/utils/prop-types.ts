export const tuple = <T extends string[]>(...args: T) => args
//todo better syntax related to tuple?
export const variantsTuple = <T extends ('solid' | 'line' | 'text')[]>(...args: T) => args
export const statusTuple = <T extends ('default' | 'hover' | 'active' | 'disabled')[]>(
  ...args: T
) => args

const buttonVariants = variantsTuple('solid', 'line', 'text')
const tabVariants = variantsTuple('solid', 'line')
const inputVariants = tuple('solid', 'line')

const buttonColors = tuple('default', 'primary', 'secondary', 'success', 'warning', 'error')

const tagColors = tuple('default', 'primary', 'secondary', 'success', 'warning', 'error')

const buttonStatus = tuple('default', 'hover', 'active', 'disabled')
const tabStatus = statusTuple('default', 'hover', 'active', 'disabled')

const messageColors = tuple('default', 'primary', 'success', 'warning', 'error')

const selectVariants = tuple('line', 'text')

const normalSizes = tuple('mini', 'small', 'medium', 'large')

const tagSizes = tuple('small', 'medium', 'large')

const normalTypes = tuple('default', 'primary', 'secondary', 'success', 'warning', 'error')

const inputColors = tuple('default', 'primary', 'success', 'warning', 'error')

const themeTypes = tuple('dark', 'light')

const snippetColors = tuple('default', 'secondary', 'success', 'warning', 'error', 'dark', 'lite')

const cardVariants = tuple('solid', 'line')

const cardColors = tuple('default', 'primary', 'success', 'warning', 'error')

const copyTypes = tuple('default', 'slient', 'prevent')

const triggerTypes = tuple('hover', 'click')

const placement = tuple(
  'top',
  'topStart',
  'topEnd',
  'left',
  'leftStart',
  'leftEnd',
  'bottom',
  'bottomStart',
  'bottomEnd',
  'right',
  'rightStart',
  'rightEnd',
)

const dividerAlign = tuple('start', 'center', 'end', 'left', 'right')

const paginationVariants = tuple('line', 'solid')

export type ButtonColors = typeof buttonColors[number]

export type TagColors = typeof tagColors[number]

export type ButtonVariants = typeof buttonVariants[number]
export type TabVarient = typeof tabVariants[number]

export type TagVariants = ButtonVariants

export type ButtonStatus = typeof buttonStatus[number]
export type TabStatus = typeof tabStatus[number]

export type MessageColors = typeof messageColors[number]

export type NormalSizes = typeof normalSizes[number]

export type TagSizes = typeof tagSizes[number]

export type SelectVariants = typeof selectVariants[number]

export type NormalTypes = typeof normalTypes[number]

export type InputColors = typeof inputColors[number]

export type InputVariantTypes = typeof inputVariants[number]

export type ThemeTypes = typeof themeTypes[number]

export type SnippetColors = typeof snippetColors[number]

export type CardColors = typeof cardColors[number]

export type CardVariants = typeof cardVariants[number]

export type CopyTypes = typeof copyTypes[number]

export type TriggerTypes = typeof triggerTypes[number]

export type Placement = typeof placement[number]

export type DividerAlign = typeof dividerAlign[number]

export type PaginationVariants = typeof paginationVariants[number]
