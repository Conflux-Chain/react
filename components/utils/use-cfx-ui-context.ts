import React from 'react'
import { ToastWithID } from '../use-toasts/toast-container'
import { MessageItemProps } from '../use-messages/message-item'
import { NotificationItemProps } from '../use-notifications/notification-item'

export type UpdateToastsFunction<T> = (fn: (toasts: Array<T>) => Array<T>) => any
export type UpdateMessagesFunction<T> = (fn: (messages: Array<T>) => Array<T>) => any
export type UpdateNotificationsFunction<T> = (fn: (notifications: Array<T>) => Array<T>) => any

export interface CfxUIContextParams {
  messages: Array<MessageItemProps>
  updateMessages: UpdateMessagesFunction<MessageItemProps>
  notifications: Array<NotificationItemProps>
  updateNotifications: UpdateNotificationsFunction<NotificationItemProps>
  toasts: Array<ToastWithID>
  toastHovering: boolean
  updateToasts: UpdateToastsFunction<ToastWithID>
  updateToastHoverStatus: Function
}

const defaultParams: CfxUIContextParams = {
  messages: [],
  updateMessages: t => t,
  notifications: [],
  updateNotifications: t => t,
  toasts: [],
  toastHovering: false,
  updateToasts: t => t,
  updateToastHoverStatus: () => {},
}

export const CfxUIContent: React.Context<CfxUIContextParams> = React.createContext<
  CfxUIContextParams
>(defaultParams)

export const useCfxUIContext = (): CfxUIContextParams =>
  React.useContext<CfxUIContextParams>(CfxUIContent)
