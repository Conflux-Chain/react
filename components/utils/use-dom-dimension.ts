import React, { useEffect, useRef, useState } from 'react'

export default function useDOMDimension<E extends HTMLElement>(
  property: 'clientWidth' | 'offsetWidth' = 'clientWidth', //more to come
  ...gurads: any[]
): [number, React.RefObject<E>] {
  const ref = useRef<E>(null)
  const [width, setWidth] = useState(0)
  useEffect(() => {
    if (ref.current) {
      setWidth(ref.current[property])
    }
  }, [...gurads])
  return [width, ref]
}
