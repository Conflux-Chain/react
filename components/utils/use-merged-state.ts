import { useRef, useState, useCallback } from 'react'
// https://github.com/react-component/util/blob/4045d568885d65a5491ab22347682f5e31da31dc/src/hooks/useMergedState.ts
export default function useControlledState<T>(
  defaultValue: T,
  option?: {
    value?: T
    onChange?: (value: T, prevValue: T) => void
    postState?: (value: T) => T
  },
): [T, (value: T) => void] {
  const { value, onChange, postState } = option || {}
  const [innerValue, setInnerValue] = useState<T>(value !== undefined ? value : defaultValue)

  // controlled take priority of uncontrolled
  let calcualteValue = value !== undefined ? value : innerValue
  if (postState) {
    calcualteValue = postState(calcualteValue)
  }

  // value inside setValue should be updated
  const mergedValue = useRef(calcualteValue)
  mergedValue.current = calcualteValue

  // do not change the identity of setValue
  const setValue = useCallback((newValue: T) => {
    if (!value) {
      console.log('setValue', newValue)
      setInnerValue(newValue)
    }
    if (mergedValue.current !== newValue && onChange) {
      onChange(newValue, mergedValue.current)
    }
  }, [])

  return [mergedValue.current, setValue]
}
