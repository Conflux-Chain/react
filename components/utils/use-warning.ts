const warningStack: { [key: string]: boolean } = {}

const useWarning = (
  message: string,
  component?: string,
  level?: 'error' | 'warn' | 'info' | 'log',
) => {
  const tag = component ? ` [${component}]` : ' '
  const log = `[@cfxjs/react-ui]${tag}: ${message}`

  if (typeof console === 'undefined') return
  if (warningStack[log]) return
  warningStack[log] = true

  if (level) return console[level](log)

  if (process.env.NODE_ENV !== 'production') {
    return console.error(log)
  }

  console.warn(log)
}

export default useWarning
