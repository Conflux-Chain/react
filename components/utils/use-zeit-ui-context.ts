import React from 'react'
import { ToastWithID } from '../use-toasts/toast-container'
import { MessageWithID } from '../use-messages/toast-container'

export type UpdateToastsFunction<T> = (fn: (toasts: Array<T>) => Array<T>) => any
export type UpdateMessagesFunction<T> = (fn: (messages: Array<T>) => Array<T>) => any

export interface ZeitUiContextParams {
  messages: Array<MessageWithID>
  updateMessages: UpdateMessagesFunction<ToastWithID>
  toasts: Array<ToastWithID>
  toastHovering: boolean
  updateToasts: UpdateToastsFunction<ToastWithID>
  updateToastHoverStatus: Function
}

const defaultParams: ZeitUiContextParams = {
  messages: [],
  updateMessages: t => t,
  toasts: [],
  toastHovering: false,
  updateToasts: t => t,
  updateToastHoverStatus: () => {},
}

export const ZEITUIContent: React.Context<ZeitUiContextParams> = React.createContext<
  ZeitUiContextParams
>(defaultParams)

export const useZEITUIContext = (): ZeitUiContextParams =>
  React.useContext<ZeitUiContextParams>(ZEITUIContent)
