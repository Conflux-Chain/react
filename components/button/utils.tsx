import React, { ReactNode } from 'react'
import { NormalSizes } from '../utils/prop-types'
import ButtonIcon from './button-icon'
import { ButtonProps } from './button'
import { ButtonGroupConfig } from '../button-group/button-group-context'

export const getButtonChildrenWithIcon = (
  loading: boolean,
  auto: boolean,
  size: NormalSizes,
  children: ReactNode,
  icons: {
    icon?: React.ReactNode
    iconRight?: React.ReactNode
  },
  isTag: boolean,
) => {
  const { icon, iconRight } = icons
  const hasIcon = icon || iconRight
  const isRight = Boolean(iconRight)
  const paddingForAutoMode =
    auto || size === 'mini' || isTag
      ? `calc(var(--zeit-ui-button-height) / 2 + var(--zeit-ui-button-padding) * .5)`
      : 0
  if (!hasIcon) return <div className={`text ${loading ? 'hidden' : ''}`}>{children}</div>
  return (
    <>
      <ButtonIcon isRight={isRight}>{hasIcon}</ButtonIcon>
      <div className={`text ${isRight ? 'right' : 'left'} ${loading ? 'hidden' : ''}`}>
        {children}
        <style jsx>{`
          .left {
            padding-left: ${paddingForAutoMode};
          }
          .right {
            padding-right: ${paddingForAutoMode};
          }
        `}</style>
      </div>
    </>
  )
}

export const filterPropsWithGroup = (
  props: React.PropsWithChildren<ButtonProps>,
  config: ButtonGroupConfig,
): ButtonProps => {
  if (!config.isButtonGroup) return props
  return {
    ...props,
    auto: true,
    shadow: false,
    ghost: config.ghost || props.ghost,
    size: config.size || props.size,
    variant: config.variant || props.variant,
    color: config.color || props.color,
    disabled: config.disabled || props.disabled,
  }
}
