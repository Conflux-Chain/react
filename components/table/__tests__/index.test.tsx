import Table from '../'
import { mount } from 'enzyme'
import React from 'react'
import { nativeEvent } from 'tests/utils'

describe('Table', () => {
  it('should render correctly', () => {
    const columns = [
      { title: 'title1', dataIndex: 'a', key: 'a', width: 100 },
      { id: '123', title: 'title2', dataIndex: 'b', key: 'b', width: 100, align: 'right' },
      { title: 'title3', dataIndex: 'c', key: 'c', width: 200 },
      {
        title: 'Operations',
        dataIndex: '',
        key: 'd',
        render(_, record) {
          return (
            <a
              onClick={e => {
                e.preventDefault()
                console.log('Operate on:', record)
              }}
              href="#">
              Operations
            </a>
          )
        },
      },
    ]
    const data = [
      { a: '123', key: '1' },
      { a: 'cdd', b: 'edd', key: '2' },
      { a: '1333', c: 'eee', d: 2, key: '3' },
    ]
    const wrapper = mount(
      <div>
        <Table columns={columns} data={data} />
        <Table variant="solid" columns={columns} data={data} />
      </div>,
    )
    expect(wrapper.html()).toMatchSnapshot()
  })
})
