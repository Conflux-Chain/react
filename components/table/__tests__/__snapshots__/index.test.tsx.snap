// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render correctly 1`] = `"<div><div><div class=\\"table mock variant-line\\"><div class=\\"table-container\\"><div class=\\"table-content\\"><table style=\\"table-layout: auto;\\"><colgroup><col style=\\"width: 100px; min-width: 100px;\\"><col style=\\"width: 100px; min-width: 100px;\\"><col style=\\"width: 200px; min-width: 200px;\\"></colgroup><thead class=\\"table-thead\\"><tr><th class=\\"table-cell\\">title1</th><th class=\\"table-cell\\" style=\\"text-align: right;\\">title2</th><th class=\\"table-cell\\">title3</th><th class=\\"table-cell\\">Operations</th></tr></thead><tbody class=\\"table-tbody\\"><tr data-row-key=\\"1\\" class=\\"table-row table-row-even table-row-level-0\\"><td class=\\"table-cell\\">123</td><td class=\\"table-cell\\" style=\\"text-align: right;\\"></td><td class=\\"table-cell\\"></td><td class=\\"table-cell\\"><a href=\\"#\\">Operations</a></td></tr><tr data-row-key=\\"2\\" class=\\"table-row table-row-level-0\\"><td class=\\"table-cell\\">cdd</td><td class=\\"table-cell\\" style=\\"text-align: right;\\">edd</td><td class=\\"table-cell\\"></td><td class=\\"table-cell\\"><a href=\\"#\\">Operations</a></td></tr><tr data-row-key=\\"3\\" class=\\"table-row table-row-even table-row-level-0\\"><td class=\\"table-cell\\">1333</td><td class=\\"table-cell\\" style=\\"text-align: right;\\"></td><td class=\\"table-cell\\">eee</td><td class=\\"table-cell\\"><a href=\\"#\\">Operations</a></td></tr></tbody></table></div></div></div></div><div><div class=\\"table mock variant-solid\\"><div class=\\"table-container\\"><div class=\\"table-content\\"><table style=\\"table-layout: auto;\\"><colgroup><col style=\\"width: 100px; min-width: 100px;\\"><col style=\\"width: 100px; min-width: 100px;\\"><col style=\\"width: 200px; min-width: 200px;\\"></colgroup><thead class=\\"table-thead\\"><tr><th class=\\"table-cell\\">title1</th><th class=\\"table-cell\\" style=\\"text-align: right;\\">title2</th><th class=\\"table-cell\\">title3</th><th class=\\"table-cell\\">Operations</th></tr></thead><tbody class=\\"table-tbody\\"><tr data-row-key=\\"1\\" class=\\"table-row table-row-even table-row-level-0\\"><td class=\\"table-cell\\">123</td><td class=\\"table-cell\\" style=\\"text-align: right;\\"></td><td class=\\"table-cell\\"></td><td class=\\"table-cell\\"><a href=\\"#\\">Operations</a></td></tr><tr data-row-key=\\"2\\" class=\\"table-row table-row-level-0\\"><td class=\\"table-cell\\">cdd</td><td class=\\"table-cell\\" style=\\"text-align: right;\\">edd</td><td class=\\"table-cell\\"></td><td class=\\"table-cell\\"><a href=\\"#\\">Operations</a></td></tr><tr data-row-key=\\"3\\" class=\\"table-row table-row-even table-row-level-0\\"><td class=\\"table-cell\\">1333</td><td class=\\"table-cell\\" style=\\"text-align: right;\\"></td><td class=\\"table-cell\\">eee</td><td class=\\"table-cell\\"><a href=\\"#\\">Operations</a></td></tr></tbody></table></div></div></div></div></div>"`;
